# Odhad závislosti mezi proměnnými pomocí $\chi^2$ testu
### V tomto příkladu zkusíme ověřit s 95% jistotou následující hypotézu:<br>
<p><em>Pokud viník nehody byl pod silným vlivem alkoholu, došlo častěji k těžkým zdravotním následkům.</em></p>

## Požadované prerekvizity z přednášek
### Teorie
 * Kontingenční tabulka
 * Chi-kvadrát test
 
### Python
 * Manipulace s daty dat (pandas)
 * Analýza statistik (scipy)

Na začátku stáhneme dataset

In [9]:
import pandas as pd

df = pd.read_pickle("accidents.pkl.gz")
df.head()

p1 p36 p37         p2a weekday(p2a)   p2b p6  p7  p8  p9  ...  l  \
0  002100160001   4      2016-01-01            5  0055  1   1   0   2  ...      
1  002100160002   4      2016-01-01            5  0130  1   3   0   2  ...      
2  002100160003   5      2016-01-01            5  0100  1   2   0   2  ...      
3  002100160004   6      2016-01-01            5  0120  9   0   0   2  ...      
4  002100160005   6      2016-01-01            5  2560  2   0   0   2  ...      

        n  o                       p       q       r       s  \
0  711403     Souhlasnýsesměremúseku  Pomalý  554782  451622   
1                                                              
2                                                              
3                                                              
4                                                              

                    t  p5a  region  
0  GN_V0.1UIR-ADR_410    1     PHA  
1                        1     PHA  
2                        1     PHA  
3                        1     PHA  
4                        1     PHA  

[5 rows x 65 columns]

Potom vyloučíme data, které nejsou potřebné pro řešení dané úlohy

In [10]:
df = df[['p11', 'p13a', 'p13b']]
df = df.loc[(df.p11 != 4) & (df.p11 != 5)]

df

p11  p13a  p13b
0         2     0     0
1         2     0     0
2         2     0     0
3         9     0     0
4         0     0     0
...     ...   ...   ...
487156    0     0     0
487157    2     0     0
487158    0     0     0
487159    2     0     0
487160    0     0     0

[485683 rows x 3 columns]

Pomocí tohoto datasetu vytvoříme kontingenční tabulku, která bude obsahovat binarizovaná data o vlivu alkoholu vůči následkům havárii 

In [11]:
df['p11'] = (df['p11'] >= 7)
df['p13'] = (df.p13a + df.p13b > 0)
df = pd.crosstab(df.p11, df.p13)

df

p13     False  True 
p11                 
False  457528  10777
True    16492    886

Pracujeme s tabulku 2x2 pro kterou vypočítáme $\chi^2$ test. Funkce pro počítání $\chi^2$ je implementovaná v scipy.stats.chi2 a hodnotu p-value tímpádem lze pro daný dataset vypočítat jako:

In [12]:
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(df)
p

2.0971505700338304e-123


Hodnota $2.09*10^{-123}$ je mnohem menší než $0.05$. Proto zamítneme nulovou hypotézu a můžeme prohlásit, že je zde statisticky významná korelace mezi silným vlivem alkoholu a těžkými zdravotními následkami.

# Závěr 

Pomocí tohoto testu jde vidět, že je zde statisticky významná korelace mezi silným vlivem alkoholu a těžkými zdravotními následkami.

**Pozor:** [Korelace neimplikuje kauzalitu](https://cs.wikipedia.org/wiki/Korelace_neimplikuje_kauzalitu). Pokud nějaké dvě proměnné spolu korelují, nelze z toho ještě vyvozovat, že jedna je příčinou a druhá kauzálním následkem.